authz/rbac_translator.go

```diff
@@ -0,0 +1,311 @@
+/*
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package authz exposes methods to manage authorization within gRPC.
+package authz
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	v3rbacpb "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v3"
+	v3routepb "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	v3matcherpb "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
+)
+
+type header struct {
+	Key    *string   `json:"key"`
+	Values *[]string `json:"values"`
+}
+
+type peer struct {
+	Principals *[]string `json:"principals"`
+}
+
+type request struct {
+	Paths   *[]string `json:"paths"`
+	Headers *[]header `json:"headers"`
+}
+
+type rule struct {
+	Name    *string  `json:"name"`
```

Comment:
Have we considered the option to make it a plain string instead of a `*string` here? Same for all other fields here..

Reply:
The reason for doing that was to differentiate between if a value is set to empty or unset. For ex. name field. Now, I feel like it does not add much value. So I reverted it back to use plain string. Similarly for other fields.